File: fsharp/life/Life.fs
open Printf
open System

let generations = 1000;;
let show_work = false;;

type coord = {x : int; y : int};;
type destiny = Live | Die ;;
type change = { coord : coord; destiny : destiny };;

let make_coord a b =
  { x = a; y = b };;

let enlarge box coord =
  let (minx, miny, maxx, maxy) = box in
  (min minx coord.x, min miny coord.y, max maxx coord.x, max maxy coord.y)
;;

let bbox lst =
  let x = Int32.MaxValue in
  let n = Int32.MinValue in
  Set.fold enlarge (x, x, n, n) lst
;;

let esc =
    Convert.ToChar 27;;

let printBoard s =
  printf "%c[2J" esc;
  printf "%c[;H" esc;
  let (minx, miny, maxx, maxy) = bbox s in
  for y = maxy downto miny do
    for x = minx to maxx do
      printf "%c" (if Set.contains (make_coord x y) s then '@' else ' ');
    done;
    printf "\n";
  done;
  System.Threading.Thread.Sleep( 1000 / 30)
  ()
;;

let make_change c d =
  { coord = c; destiny = d };;

let live c =
  make_change c Live

let die c =
  make_change c Die

let is destiny change =
    (compare change.destiny destiny) = 0
;;

let toLiveSet changeSet =
  Set.ofSeq (Set.map (fun x -> x.coord) changeSet)
;;

let applyChanges liveSet changes =
  let toLive = toLiveSet (Set.filter (is Live) changes) in
  let toDie = toLiveSet (Set.filter (is Die) changes) in
  Set.union toLive (Set.difference liveSet toDie)
;;

let eight coord =
  List.map (fun (x, y) -> { x = coord.x + x; y = coord.y + y }) [
      (-1,-1); (-1,0); (-1,1);
      (0,-1); (0, 1);
      (1,-1); (1, 0); (1, 1)];;

let computeNeighbors changes =
  let possible =
    Set.map (fun change -> change.coord) changes in
  Set.ofList(List.concat(List.map eight (Set.toList possible)));;

let neighborCount liveSet coord =
  List.length (List.filter (fun c -> Set.contains c liveSet) (eight coord));;

let computeChange liveSet coord =
  match neighborCount liveSet coord with
  | 2 -> None
  | 3 -> if (Set.contains coord liveSet) then None else Some (live coord)
  | _ -> if (Set.contains coord liveSet) then Some (die coord) else None
;;

let computeChanges liveSet neighbors =
  Set.ofList
        (List.choose (computeChange liveSet) (Set.toList neighbors))
;;

let rec run count liveSet changes =
  match count with
  | 0 -> ()
  | _ ->
     let updated = applyChanges liveSet changes in
     if show_work then
       printBoard updated
     else
       ();
     let neighbors = computeNeighbors changes in
     let nextGen = computeChanges updated neighbors in
     run (count - 1) updated nextGen
;;

let toLive xy =
  let (x, y) = xy in
  live (make_coord x y)
;;

let run1 ignored =
  let r_pentomino = [(0, 0); (0, 1); (1, 1); (-1, 0); (0, -1)] in
  let startChanges = Set.ofList (List.map toLive r_pentomino) in
  run generations (Set.empty) startChanges;;

let time f =
  let start = System.DateTime.Now in
  let res = f () in
  let diff = System.DateTime.Now.Subtract start in
  (res, diff.TotalMilliseconds / 1000.0)
;;

let main =
  for i = 1 to (if show_work then 1 else 5) do
    let (res, diff) = time run1 in
    printf "%.4f generations per sec\n" ((float generations) / diff)
  done
;;

File: csharp/life/Life.cs
using System;
using System.Collections.Generic;
// I tried IImmutableSet/ImmutableHash map and it was 3x slower!
// using System.Collections.Immutable;
using System.Linq;

namespace life
{
    struct Coord
    {
        public readonly int x;
        public readonly int y;
        public Coord(int x, int y)
        {
            this.x = x;
            this.y = y;
        }
    }
    enum Destiny { Live, Die }

    struct Change
    {
        public readonly Coord coord;
        public readonly Destiny destiny;
        public Change(Coord coord, Destiny destiny)
        {
            this.coord = coord;
            this.destiny = destiny;
        }
    }

    struct Box
    {
        public readonly Coord lowerLeft;
        public readonly Coord upperRight;
        public Box(Coord lowerLeft, Coord upperRight)
        {
            this.lowerLeft = lowerLeft;
            this.upperRight = upperRight;
        }
    }

    class Life
    {
        static readonly int GENERATIONS = 1000;
        static readonly bool SHOW_WORK = false;

        public static readonly Coord[] rPentomino = new Coord[] {
                new Coord(0, 0),
                new Coord(0, 1),
                new Coord(1, 1),
                new Coord(-1, 0),
                new Coord(0, -1)
                };

        static ISet<Coord> ApplyUpdates(
            ISet<Coord> liveSet,
            IEnumerable<Change> changes)
        {
            var toDie =
                from c in changes where c.destiny == Destiny.Die select c.coord
                ;
            var toLive =
                from c in changes where c.destiny == Destiny.Live select c.coord
                ;
            var result = new HashSet<Coord>(liveSet);
            result.ExceptWith(toDie);
            result.UnionWith(toLive);
            return result;
        }

        private static readonly Box tinyBox = new(
            new Coord(int.MaxValue, int.MaxValue),
            new Coord(int.MinValue, int.MinValue)
            );

        private static Box Enlarge(Box box, Coord c)
        {
            return new(
                new Coord(Math.Min(box.lowerLeft.x, c.x), Math.Min(box.lowerLeft.y, c.y)),
                new Coord(Math.Max(box.upperRight.x, c.x), Math.Max(box.upperRight.y, c.y))
                );
        }

        private static void PrintLiveSet(ISet<Coord> liveSet)
        {
            Console.Write("\x1b[2J\x1b[;H");
            if (liveSet.Count == 0)
            {
                return;
            }
            var boundingBox = liveSet.Aggregate(tinyBox, Enlarge);
            var ur = boundingBox.upperRight;
            var ll = boundingBox.lowerLeft;
            foreach (int y in Enumerable.Range(ll.y, ur.y - ll.y + 1).Reverse())
            {
                foreach (int x in Enumerable.Range(ll.x, ur.x - ll.x + 1))
                {
                    if (liveSet.Contains(new Coord(x, y)))
                    {
                        Console.Write("@");
                    }
                    else
                    {
                        Console.Write(" ");
                    }
                }
                Console.WriteLine("");
            }
            System.Threading.Thread.Sleep(1000 / 30);
        }

        private static readonly int[] offsets = { -1, 0, 1 };

        private static IEnumerable<Coord> Eight(Coord c)
        {
            var result = new List<Coord>(8);
            foreach (var y in offsets)
            {
                foreach (var x in offsets)
                {
                    if (x != 0 || y != 0)
                    {
                        result.Add(new Coord(c.x + x, c.y + y));
                    }
                }
            }
            return result;
        }

        private static ISet<Coord> ComputeNeighbors(IEnumerable<Change> changes)
        {
            return (
                from c in changes
                select Eight(c.coord)
            )
            .SelectMany(x => x)
            .ToHashSet();
        }

        private static int NeighborCount(ISet<Coord> liveSet, Coord n)
        {
            return (from c in Eight(n) where liveSet.Contains(c) select 1).Sum();
        }

        private static IEnumerable<Change> ComputeUpdates(ISet<Coord> liveSet, ISet<Coord> neighbors)
        {
            var result = new List<Change>(neighbors.Count);
            foreach (var n in neighbors)
            {
                switch (NeighborCount(liveSet, n))
                {
                    case 2:
                        break;
                    case 3:
                        if (!liveSet.Contains(n))
                        {
                            result.Add(new Change(n, Destiny.Live));
                        }
                        break;
                    default:
                        if (liveSet.Contains(n))
                        {
                            result.Add(new Change(n, Destiny.Die));
                        }
                        break;
                }
            }
            return result;
        }

        private static void Run()
        {
            var updates = from c in rPentomino
                          select new Change(c, Destiny.Live);
            ISet<Coord> liveSet = new HashSet<Coord>();

            foreach (int generation in Enumerable.Range(1, GENERATIONS))
            {
                liveSet = ApplyUpdates(liveSet, updates);
                if (SHOW_WORK)
                {
                    PrintLiveSet(liveSet);
                }
                var neighbors = ComputeNeighbors(updates);
                updates = ComputeUpdates(liveSet, neighbors);
            }
        }

        static void Main(string[] _)
        {
            if (SHOW_WORK)
            {
                Run();
            }
            else
            {
                foreach (int _unused in Enumerable.Range(0, 5))
                {
                    var start = DateTime.Now;
                    Run();
                    var stop = DateTime.Now;
                    var diffMillis = stop.Subtract(start).TotalMilliseconds;
                    Console.WriteLine(String.Format("{0} generations per sec", GENERATIONS * 1000 / diffMillis));
                }
            }
        }
    }
}
